python/cudf/cudf/core/column/numerical.py

```diff
@@ -327,6 +327,8 @@ def find_and_replace(self, to_replace, replacement, all_nan):
         output = libcudf.replace.replace(
             replaced, to_replace_col, replacement_col
         )
+        if not (self.dtype.kind in "i" and output.dtype.kind in "f"):
+            output = output.astype(self.dtype)
```

Comment:
Is the `libcudf.replace.replace` call returning a different dtype than the input column? Why do we need to typecast here?

Reply:
The `libcudf.replace` call itself works as expected with types. The issue is while creating `to_replace_col` and `replacement_col` from scalars or lists the default type is going to be `int64/float64` which when normalized gives an `int64/float64` output.

The simpler approach would be to use the original column dtype while creating those two columns (at the beginning of this method). 

The reason I had to use this approach instead is to account for corner cases like this:
```
a = cudf.Series([1,2,3],dtype="int8")
a.replace(257,12)
```
Pandas return `[1,2,3]` as the output with type "int8".

By casting the inputs `(257 and 12)` to int8 cudf would treat 257 as 1 and return `[12,2,3]` as the output.

Reply:
So the libcudf function under the hood will throw an error if the dtypes of the input and replacement columns don't match. If we could work out logic to allow the replacement column to only be upcasted, that would be the ideal solution from my perspective. This way if someone tries to pass an int16 or int32 value to an int8 column it will loudly fail, but if they pass an int8 value to an int32 column it would gracefully upcast and succeed.

Reply:
> So the libcudf function under the hood will throw an error if the dtypes of the input and replacement columns don't match.

Yes 
> If we could work out logic to allow the replacement column to only be upcasted, that would be the ideal solution from my perspective. 

This gets slightly tricky because this function can be called with scalars and python lists in which case the default type of the replacement column is an `int64/float64`. It would then some logic like the helper function I had implemented to find the min/max of the replacement list and determine which dtype these boundary values could be represented by. 
> This way if someone tries to pass an int16 or int32 value to an int8 column it will loudly fail, but if they pass an int8 value to an int32 column it would gracefully upcast and succeed.

We can then have logic checks if the replacement columns dtype > existing columns dtype we might choose to scalar overflow or throw an error (either works). And if the existing columns dtype > replacement columns dtype everything gets normalized/upcasted to the existing columns dtype (with the exception of when the scalar is a float).

Since there are multiple cases I am creating a table and it would be clear if we could decide on the behavior we expect based on the condition.

| Input col dtype | to_replace value | Replacement value | Expected output dtype |
|-----------------|------------------|-------------------|-----------------------|
| int8            | int64            | int64             |                       |
| int8            | int64            | float64           |                       |
| int8            | float64          | int64             |                       |
| float32         | float64          | float64           |                       |
| float32         | float64          | float32           |                       |

Sorry for making this more complex than it might seem to you

Reply:
> This gets slightly tricky because this function can be called with scalars and python lists in which case the default type of the replacement column is an int64/float64. It would then need some logic like the helper function I had implemented to find the min/max of the replacement list and determine which dtype these boundary values could be represented by.

Yes we'll need that function back, sorry 😓. Ideally we can cast python scalars / lists to their minimum viable dtype and handle it from there. If the user passes in a strongly typed object such as a numpy array / pyarrow array then we use the type of the array.

Table below assumes that we convert Python scalars / lists to their minimum viable dtype:

| Input col dtype | to_replace value | Replacement value | Expected output dtype |
| --- | --- | --- | --- |
| int8 | int8 | int8 | int8 |
| int8 | int16 | int8 | error |
| int8 | int8 | int16 | error |
| int16 | int16 | int16 | int16 |
| int16 | int8 | int8 | int16 |
| int16 | int8 | int16 | int16 |
| int16 | int16 | int8 | int16 |
| float32 | float32 | float32 | float32 |
| float32 | float32 | float64 | error |
| float32 | float64 | float32 | error |
| float32 | float64 | float64 | error |
| float32 | int8 | int8 | float32 |
| float32 | int32 | int32 | error |
| float64 | float64 | float64 | float64 |
| float64 | float32 | float32 | float64 |
| float64 | float32 | float64 | float64 |
| float64 | float64 | float32 | float64 |
| float64 | int8 | int8 | float64 |
| float64 | int64 | int64 | float64 |

The logic is basically that if `to_replace_value.dtype` and `replacement_value.dtype` <= `input_col.dtype` then `output_col.dtype` = `input_col.dtype`, otherwise, throw an exception

Reply:
Numpy dtypes allow for such comparisons:
```python
np.dtype('int64') > np.dtype('float64')
# False

np.dtype('float64') > np.dtype('int64')
# True
```

Reply:
Thanks. I was working on the implementation and ran into another case. What if the original column has integer dtype (int8/16/32/64) and the replacement col has a floating point dtype (f32/64). Based on the discussion above, the implementation I have prototyped would through an error. Note pandas allows this and would upcast the original column to float (both in the case of `series.replace` and `series.where`). Wanted to reclarify before committing to a relatively dirty commit history for this pr.

Reply:
We should throw an error if the original col is integer and the replacement is float.